[spmd] expose input_batch_dim to DataParallelMode

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99964
* #99901
* #99900
* __->__ #99899
* #99898
* #99374
* #99489
* #99436
* #99897

This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

wanchaol committed:
[spmd] expose input_batch_dim to DataParallelMode

This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

[ghstack-poisoned]

awgu reviewed (approved):
LGTM.

I wonder if the user passes in an incorrect `input_batch_dim`, how will the error propagate and manifest?

wanchaol committed:
Update on "[spmd] expose input_batch_dim to DataParallelMode"


This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

[ghstack-poisoned]

wanchaol commented:
> I wonder if the user passes in an incorrect `input_batch_dim`, how will the error propagate and manifest?

Yeah we definitely should harden this, basically we now would treat the user passed in `input_batch_dim` as source of truth, if that's incorrect, we probably need some logic to infer it does not align with what we observed in the graph, and we error out to the user. Not sure how this could be easily done yet, I'll add some TODO so that we won't forget it

wanchaol committed:
Update on "[spmd] expose input_batch_dim to DataParallelMode"


This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] expose input_batch_dim to DataParallelMode"


This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] expose input_batch_dim to DataParallelMode"


This PR exposes the input batch dim to the DataParallelMode so that
we could have explicit control of which input dim is batch dim

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -60,6 +60,7 @@ class DataParallel(ParallelMode):
     def __init__(
         self,
         parallel_style: str = "replicate",
+        input_batch_dim: int = 0,
```
do we need a test to cover the case where batch dim is not dim0?

wanchaol replied:
Yep good point, let me try to add a test case in a separate PR to avoid trigger CI again
